alibi_detect/od/pytorch/base.py

```diff
@@ -0,0 +1,239 @@
+
+        Parameters
+        ----------
+        x
+            Data to aggregate and normalize.
+
+        Returns
+        -------
+        `torch.Tensor` or just returns original data
+        """
+        if hasattr(self, 'ensembler') and self.ensembler is not None:
+            # `type: ignore` here becuase self.ensembler here causes an error with mypy when using torch.jit.script.
+            # For some reason it thinks self.ensembler is a torch.Tensor and therefore is not callable.
+            return self.ensembler(x)  # type: ignore
+        else:
+            return x
+
+    def _classify_outlier(self, scores: torch.Tensor) -> torch.Tensor:
+        """Classify the data as outlier or not.
+
+        Parameters
+        ----------
+        scores
+            Scores to classify. Larger scores indicate more likely outliers.
+
+        Returns
+        -------
+        `torch.Tensor` or ``None``
+        """
+        return scores > self.threshold if self.threshold_inferred else None
+
+    def _p_vals(self, scores: torch.Tensor) -> torch.Tensor:
+        """Compute p-values for the scores.
+
+        Parameters
+        ----------
+        scores
+            Scores to compute p-values for.
+
+        Returns
+        -------
+        `torch.Tensor` or ``None``
+        """
+        return (1 + (scores[:, None] < self.val_scores).sum(-1))/len(self.val_scores) \
+            if self.threshold_inferred else None
+
+    def infer_threshold(self, x: torch.Tensor, fpr: float) -> None:
+        """Infer the threshold for the data. Prerequisite for outlier predictions.
+
+        Parameters
+        ----------
+        x
+            Data to infer the threshold for.
+        fpr
+            False positive rate to use for threshold inference.
+
+        Raises
+        ------
+        ValueError
+            Raised if `fpr` is not in ``(0, 1)``.
+        """
+        if not 0 < fpr < 1:
+            ValueError('`fpr` must be in `(0, 1)`.')
+        self.val_scores = self.score(x)
+        self.val_scores = self._ensembler(self.val_scores)
+        self.threshold = torch.quantile(self.val_scores, 1-fpr)
+        self.threshold_inferred = True
+
+    def predict(self, x: torch.Tensor) -> TorchOutlierDetectorOutput:
+        """Predict outlier labels for the data.
+
+        Computes the outlier scores. If the detector is not fit on reference data we raise an error.
+        If the threshold is inferred, the outlier labels and p-values are also computed and returned.
+        Otherwise, the outlier labels and p-values are set to ``None``.
+
+        Parameters
+        ----------
+        x
+            Data to predict.
+
+        Raises
+        ------
+        ValueError
+            Raised if the detector is not fit on reference data.
+
+        Returns
+        -------
+        `TorchOutlierDetectorOutput`
+            Output of the outlier detector.
```

Comment:
Is this `Returns` description valid syntax? Does it render OK?

Reply:
I think it's [correct](https://numpydoc.readthedocs.io/en/stable/format.html#returns). It renders alright but I think it should be a link to the type so will make that change.

Reply:
Mmmn fair enough. I'm not sure we need to put the type info there though, since we have the type info in the function/method signature and this is parsed by [sphinx-autodoc-typehints](https://github.com/tox-dev/sphinx-autodoc-typehints)? i.e. for `Parameters` we have:

```python
Parameters
----------
name
     description
```

rather than:

```python
Parameters
----------
name : type 
      description
```

To be fair there are a few places in `alibi`/`alibi-detect` where we do include the parameter name in `Returns` e.g. [here](https://github.com/SeldonIO/alibi/blob/9ec8658a77694b0dc0b804f6740dd698a6707826/alibi/explainers/cfrl_tabular.py#L316-L323). But on the other hand, it doesn't seem to be done for most of this PR?